### ISSUE TITLE
ETA on languageVersion 2.0 support

### ISSUE THREAD

**Issue by wouterroos** [microsoft/vscode-azurearmtools#1669]
Hi,

Are there any plans to support languageVersion 2.0 in the VSCode tooling? 

Thanks,

**Comment by alex-frankel** [comment]
I don't believe there are any plans to support the new language version in ARM Tools. Can you help us understand what you are trying to do than you cannot accomplish with the bicep extension on the original bicep files?

**Comment by Neutrino-Sunset** [comment]
I'm not using Bicep files.

If you've effectively deprecated the Azure Resource Manager Tools by deciding not to support key new features, then it would be useful if your documentation flipping said so!

https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/overview

**Comment by alex-frankel** [comment]
Agreed that we need to document this better. There is a distinction between not supporting current features (which I would call deprecated) and not supporting new capabilities.

@puicchan / @mumian / @stephaniezyen - can we update the documentation to capture this level of detail?

**Comment by jordanmills** [comment]
Wait, so this extension is just going to ignore anything new about ARM templates?

**Comment by alex-frankel** [comment]
Any new features being added to the ARM template language are entirely to support new bicep features. There are no new features being added to ARM Templates otherwise. We strongly encourage anyone using ARM Templates to migrate to Bicep. If there are blocking issues preventing that migration, we would love to better understand those.

**Comment by jordanmills** [comment]
That is horrible and must be reconsidered.  I've found many cases where working with ARM templates is preferable or necessary and having syntax parsing, highlighting, and other support components for current versions, especially when they are clearly across Microsoft's docs, is critical.

**Comment by alex-frankel** [comment]
Do you happen to have some of those cases handy? Like I said, we really would want to understand any scenarios that are preventing migration. Would you be up for a call to discuss this in more detail?

Are there features in language version 2.0 you are looking to take advantage of? All of the current features in this extension that you are using will still continue to be supported.

**Comment by Neutrino-Sunset** [comment]
For starters Azure Template Specs don't support Bicep.

**Comment by alex-frankel** [comment]
The template spec CLI tools all support bicep for creating/updating specs with bicep files. The ARM Template is still what you would see in the portal, but at that point it is more of a deployment artifact rather than a file you should be editing.

**Comment by Neutrino-Sunset** [comment]
The ARM template is what you would see in the Portal, and it is what you would get back if you export it too, and AFAIK no way to convert that back to the Bicep file you originally had. So if you use ARM files you can use a Template Spec as a repository for your deployments, whereas if you use Bicep you have to have some other process to manage them.

Looking at the MS IAAC Azure offerings I see:

1) The ARM tooling doesn't support language version 2 and isn't planned to be updated anymore
2) Bicep isn't supported in Template Specs and you can't convert ARM back to your original Bicep
3) Neither ARM nor Bicep support destroying the resources they actually created, both only support the teardown of an entire Resource Group

MS still doesn't seem to have anything that can do what Terraform does.

**Comment by jordanmills** [comment]
> Do you happen to have some of those cases handy? Like I said, we really would want to understand any scenarios that are preventing migration. Would you be up for a call to discuss this in more detail?
> 
> Are there features in language version 2.0 you are looking to take advantage of? All of the current features in this extension that you are using will still continue to be supported.

One of mine is better stated above, and ties into the next one.  ARM templates are native, and tooling to manage and troubleshoot them is critical.  The tools to convert bicep to arm and back are far from perfect.  They have quirks around function preference that lead to really awkward conversions.  They can also take a valid ARM template, convert it to valid bicep, then convert it back into an invalid ARM template (I saw that one as recently as two or three days ago).

Retraining, especially over artificially forced obsolescence, will generate significant costs and ill will.  We have to get developers and admins to do more training/learning on something everyone believed they already had down, which is going to have direct costs and indirect costs on taking them away from billable work.  

Another big concern is invalidation of existing investments around infrastructure as code and CI/CD.  Everything is already in ARM templates.  But some genius is eventually going to decide that everything should be converted to bicep and kept in bicep, so there will be non-trivial costs in conversion and maintenance, that may or may not be realized in decreased costs for eliminating knowledge of ARM templates as a hiring requirement.  But you'll still need staff familiar with ARM templates for extraction, conversion validation, and troubleshooting.  So that's just a cost sink.

So what it looks like, especially at an executive or accounting perspective, is an ongoing march of post-investment costs for Azure.  If we're going to keep paying like this, why not just pay it to move to AWS where things seem a lot more stable.  If Microsoft does not think its own technology is worth ongoing investment, why should we invest in it?

To answer your question, the feature that seems most valuable is symbolic resource names.  And yes I you think that will help.

**Comment by alex-frankel** [comment]
To clarify, there is no explicit push to migrate if you are happy with ARM Templates as they are today. The current functionality will still be supported. We completely understand that a migration to a new codebase is a non-zero cost both in the actual migration of code and re-education/training. 

If there is more interest than expected in supporting symbolic name schema then we will need to take a fresh look at that and reconsider support for Lang Version 2.0. Thanks for clarifying that @jordanmills. @Neutrino-Sunset - is that also the functionality you are looking for?

Just as a word of caution, we would strongly recommend against treating Template Specs as a code repository. It is designed to be an artifact repository like NuGet or NPM. The source code for those template specs should still be in code repos like GitHub where you can branch, create PRs, look at commit history and diffs, etc. In that case, if you have converted to Bicep the source will always be there.

Lastly, regarding this statement @Neutrino-Sunset:

>Neither ARM nor Bicep support destroying the resources they actually created, both only support the teardown of an entire Resource Group

We have recently introduced [Deployment Stacks](https://learn.microsoft.com/azure/azure-resource-manager/bicep/deployment-stacks?tabs=azure-powershell) for exactly this purpose. It is in Public Preview. Please check it out and let us know if you it is meeting your expectations. It does not require any changes to your Template code, but does require you change the CLI command you use to deploy. Deploying as a Stack is also supported from the portal.

**Comment by DoubleCouponDay** [comment]
> 2. Bicep isn't supported in Template Specs and you can't convert ARM back to your original Bicep.

try using the Azure CLI to convert it.

`az bicep decompile --file arm.json`

**Comment by mcdonamw** [comment]
I'm just now starting to learn ARM templates and I'm starting from the templates created in the portal.  Trying to use them in VSCode with the ARM extension has led to warnings and me to researching how to get around them only to find this issue and no intention to support ARM any longer.  Just my luck.  I am using the Portal templates to learn all this stuff but now I have to re learn this stuff from the ground up or deal with conversions to Bicep on top of my initial learning as well?  That stinks.

**Comment by cf-rdegregory** [comment]
> Any new features being added to the ARM template language are entirely to support new bicep features. There are no new features being added to ARM Templates otherwise. We strongly encourage anyone using ARM Templates to migrate to Bicep. If there are blocking issues preventing that migration, we would love to better understand those.

One of the continued (required) use cases for ARM Templates are for defining Azure Marketplace Offers, where publishing a Bicep Template is simply [not an option](https://learn.microsoft.com/en-us/azure/azure-resource-manager/managed-applications/publish-service-catalog-app?tabs=azure-powershell):

> You can use Bicep to develop a managed application definition but it must be converted to ARM template JSON before you can publish the definition in Azure

However, it **also** appears that the Marketplace doesn't support the `languageVersion: 2.0` features of ARM JSON, making it impossible for me to use a compiled version of my Bicep Template in my Marketplace Offer:

![image](https://github.com/user-attachments/assets/6a05207a-25d5-4140-9f0d-2555f577ea1d)

The error is related to the `resources` property being an object (of `#/definitions/resourcesWithSymbolicName` objects), rather than an array (of `#/definitions/resourcesWithoutSymbolicName` objects), which is a `languageVersion: 2.0` feature. See the [deploymentTemplate.json schema](https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#).

---

@alex-frankel, are there any workarounds for someone in this position, aside from manually converting my 8400-line ARM Template to remove all `languageVersion: 2.0` features? 😃.

**Comment by alex-frankel** [comment]
I believe the Marketplace Offers will start supporting Language Version 2.0 and Bicep sometime in the next semester of work. Let me get you more details on that one.

**Comment by mrsmitty** [comment]
Hi @alex-frankel, we're also attempting to publish a managed application to the marketplace and come across the issue with resources needing to be an array i.e. `#/definitions/resourcesWithoutSymbolicName`

Any update on when Partner Center will support Language Version 2.0?

Thanks!

**Comment by Kaloszer** [comment]
@alex-frankel we are also looking into Azure Marketplace Offers and I just found out that Language Version 2.0 is not supported - is there any ETA on that?

**Comment by EvrenGursoy** [comment]
> I believe the Marketplace Offers will start supporting Language Version 2.0 and Bicep sometime in the next semester of work. Let me get you more details on that one.

Any update on this one @alex-frankel , we are being severely impacted by this? 

It is a wonder how a company is supposed to publish a managed app these days. After weeks making sure our bicep templates are up-to-date and following good conventions, now we need to figure out how to hack things to support language version 1.0. It was really despairing to, after waiting for @secure() outputs, having to strip those out and do all sorts of funky gymnastics with key vault. And now we have to work around not being able to use "existing" and such, and have to strip out our custom functions. It is simply not tenable. :(

**Comment by oj2m** [comment]
One of the core requirements for microsoft certification is to understand and write ARM code. While Bicep is preferred language no doubt, the fact is this tool isn't just for professionals but also a much needed resource to learning. Countless documents point to this tool to help generate consistent and reliable templates. In fact it's entirely microsoft prerogative to change their testing (they've done so before) to reflect 2.0 and update none of the learning material.

**Comment by alex-frankel** [comment]
Unfortunately, the team that owns the Marketplace contribution tooling has decided not to implement support for Bicep/LV2.0 for the time being. It is something they have decided to revisit in future planning. I will share this thread with them to make sure they've seen it.

**Comment by lswdowcoombs** [comment]
The lack of support for languageVersion 2.0 is baffling to us, and we've encountered a bit of a catch-22. Let me paint a picture that spans bicep, the marketplace, and partner center:

As part of our deployment process, we construct a dynamic number of notification hubs that support browser, android, ios, and windows-based notifications. This list is dynamic based on the number of app brands we are building and deploying for. For browser-based push notifications where a VAPID keypair is required, we must store the private and public keys in key vault for reusability across deployments and re-deployments when making infrastructure changes, which is all well and good.

The problem begins with the public key living in key vault. We use a function app to handle push notifications, and it must know what the public key for each of our app brands is, so that key for each brand is stored in the function app's environment variables. This is populated in bicep without issue when spinning up the function app and is handled by a for-loop in a module.

But due to the complexity of this, we pass the proposed app settings up from that module to the caller, so that we can merge those app settings with the existing app settings on the function app. This works, but we have to disable a linter warning in order to compile mainTemplate.json: `#disable-next-line outputs-should-not-contain-secrets`, because that public key is included in the output object. It's misleading, because this isn't necessarily a secret value, as it's available in the function app's environment variables.

We tried to use the `secure()` decorator for the output, but this is only available in later versions of bicep - versions that will generate `languageVersion 2.0` and prevent uploading the zip package in the partner center. As a result, we're stuck at version 0.25.3 until languageVersion 2.0 is supported.

The reason we've encountered this in the first place is because we currently have a private offer in the marketplace that works just fine, but in trying to create a public offer, the publish process fails at the certification step because of arm-ttk being run automatically during that process, which ignores the aforementioned linter, and spits it out as an error, preventing us from publishing a public offer of the exact same mainTemplate.json and createUiDefinition files.

With a fair amount of refactoring, we may have overcome this issue, but the number of hoops to jump through to get this to work is mind boggling. And the fact that this has not been implemented for many years is disheartening, to say the least. We hope this can be moved higher up the priority list, as we can't be the only ones waiting around for this.
